### PR TITLE
Workaround for discover themes from plugins

### LINKED ISSUE #3270: `appearance -> theme` validation warning when starting napari
## 🐛 Bug

PR #3034 introduced hook spec to add custom themes which are dynamically loaded at app startup. 
If the last active theme is not one of the defaults (dark/light) and is provided by the hook spec and napari is closed, the last used value is exported to settings file. Next time napari is started, that theme is not loaded yet so a warning is emitted and napari reverts back to safe default.

## To Reproduce

(this is not a great example because the theme provided by `new_theme` example is not provided by the hook spec)

1. Run the `examples/new_theme.py` example.
2. Close the window.
3. Re-run the example. Warning will be shown.

image

## Expected behavior

The validator should take into account non-default themes into account and not emit this warning.
The list of themes provided by `available_themes` should populated before settings are validated.

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR implements a simple workaround for problems with discovered themes from plugins

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

https://napari.zulipchat.com/#narrow/stream/212875-general/topic/theme.20plugin.20not.20working

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by brisvag** [comment]
Nice, thanks for addressing this!

 For reference, this is to solve [this issue](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/theme.20plugin.20not.20working), which is apparently related to #4948 and #4313. Did you take a look at these @Czaki?

Also, does this re-discovery risk discovering themes twice?

**Comment by Czaki** [comment]
I need to deeper investigate it.

**Comment by Czaki** [comment]
@brisvag I have updated the code to discover only this plugin is required to load the theme.

File: napari/components/viewer_model.py [hunk 1]
```diff
@@ -234,7 +234,7 @@ def _update_viewer_grid(self):
     @validator('theme')
     def _valid_theme(cls, v):
         themes = available_themes()
-        if v not in available_themes():
+        if not is_theme_available(v):
             raise ValueError(
                 trans._(
                     "Theme '{theme_name}' not found; options are {themes}.",
```
Comment (brisvag) [inline napari/components/viewer_model.py]: Probably not a big deal, but the call to `available_themes()` can be moved to after the if now.

File: napari/settings/_fields.py [hunk 2]
```diff
@@ -31,7 +31,7 @@ def validate(cls, v):
 
         value = v.lower()
         themes = available_themes()
-        if value not in available_themes():
+        if not is_theme_available(value):
             raise ValueError(
```
Comment (brisvag) [inline napari/settings/_fields.py]: Same here.

File: napari/utils/theme.py [hunk 3]
```diff
@@ -274,6 +281,35 @@ def available_themes():
...
+    if name not in _themes and _theme_path(name).exists():
+        plugin_name_file = _theme_path(name) / PLUGIN_FILE_NAME
+        print("a")
+        if not plugin_name_file.exists():
+            return False
+        print("b")
+        plugin_name = plugin_name_file.read_text()
+        print(plugin_name)
+        npe2.PluginManager.instance().register(plugin_name)
+        _install_npe2_themes(_themes)
```
Comment (brisvag) [inline napari/utils/theme.py]: some leftover prints!

File: napari/utils/theme.py [hunk 4]
```diff
@@ -317,20 +353,19 @@ def rebuild_theme_settings():
...
-        # `theme.type` is dark/light and supplies defaults for keys that
-        # are not provided by the plugin
-        d = _themes[theme.type].dict()
-        d.update(theme.colors.dict(exclude_unset=True))
-        register_theme(theme.id, d)
+    for mf in npe2.PluginManager.instance().iter_manifests(disabled=False):
+        for theme in mf.contributions.themes or ():
+            d = _themes[theme.type].dict()
+            d.update(theme.colors.dict(exclude_unset=True))
+            register_theme(theme.id, d, mf.name)
```
Comment (brisvag) [inline napari/utils/theme.py]: A bit hard to follow, could you use more verbose variable names?

**Review by brisvag — COMMENTED** [review]
Nice! Is it possible to also add a test that fails on main but works here? Because the tests so far were unable to catch the issue.

File: napari/utils/theme.py (hunk 3 above)
Comment (Czaki) [reply napari/utils/theme.py]: removed

File: napari/settings/_fields.py (hunk 2 above)
Comment (Czaki) [reply napari/settings/_fields.py]: fixed

File: napari/components/viewer_model.py (hunk 1 above)
Comment (Czaki) [reply napari/components/viewer_model.py]: fixed

File: napari/utils/theme.py (hunk 4 above)
Comment (Czaki) [reply napari/utils/theme.py]: could you check now?

File: napari/resources/_icons.py [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 ICON_PATH = (Path(__file__).parent / 'icons').resolve()
 ICONS = {x.stem: str(x) for x in ICON_PATH.iterdir() if x.suffix == '.svg'}
+PLUGIN_FILE_NAME = "plugin.txt"
```
Comment (brisvag) [inline napari/resources/_icons.py:12]: Can you elaborate a bit on this file? I'm not sure I follow what it will contain, and why it's a pure "txt".

File: napari/components/viewer_model.py [hunk 6]
```diff
@@ -233,14 +233,13 @@ def _update_viewer_grid(self):
...
-        themes = available_themes()
-        if v not in available_themes():
+        if not is_theme_available(v):
             raise ValueError(
                 trans._(
                     "Theme '{theme_name}' not found; options are {themes}.",
                     deferred=True,
                     theme_name=v,
-                    themes=themes,
+                    themes=available_themes(),
```
Comment (brisvag) [inline napari/components/viewer_model.py]:
```suggestion
                    themes=", ".join(available_themes()),
```

File: napari/resources/_icons.py (hunk 5 above)
Comment (Czaki) [reply napari/resources/_icons.py:12]: Name of plugi from which theme is obtained. 

To perform discovery only on this plugin, not all installed.

Comment (brisvag) [reply napari/resources/_icons.py:12]: Ah, I see; and how does the persistence of this file look like? Does it stick around across versions? Can/will it be deleted? What happens if one changes the name of the theme?

Comment (Czaki) [reply napari/resources/_icons.py:12]: In current napari the cache dir is separate per napari version.
There is no code for deleting the old cache. 

So this file will not be deleted but is visible in only one napari version. 

If the theme name is changed, then the old directory with SVG files and this file will still be present. But if someone uses an old name, then plugin discovery will fail and will be no added theme under this name.

Comment (brisvag) [reply napari/resources/_icons.py:12]: Ok, so there is an issue if two plugins provide a theme with the same name?

Comment (Czaki) [reply napari/resources/_icons.py:12]: Same as before this PR.

Comment (brisvag) [reply napari/resources/_icons.py:12]: Ok, then we can leave this to a followup pr!

**Review by brisvag — APPROVED** [review]

**Comment by andy-sweet** [comment]
@czaki: Does the perf test fail consistently?

After a quick look, it looks like a classic flaky test since it's waiting 100ms for something to happen. Based on the [somewhat helpful comment at the top](https://github.com/napari/napari/blob/09f2d14a8941be7da0e044389ba1c50ba602ca13/napari/_qt/perf/_tests/test_perf.py#L13), my guess is that's it's waiting for the viewer window to be shown. When running the perfmon tools some extra things happen (e.g. reading the config file, creating a dock widget to show slow events), so maybe the changes here add a little extra time to startup?

We could try patching the test to avoid doing so much and avoid flakiness. Or maybe testing less with a unit test. I think I'd also support removing the test since this is a profiling tool (though quite a useful one!).

**Comment by Czaki** [comment]
it fails if I restart the test. I will add additional commit to check it.

**Comment by Czaki** [comment]
Ok. It looks like an empty commit solves it. 
@brisvag I have made a few changes. Could you recheck it?

**Review by brisvag — APPROVED** [review]

**Review by brisvag — CHANGES_REQUESTED** [review]
Actually, I still can't seem to get my custom theme plugin to show up :/

**Comment by brisvag** [comment]
It's an empty plugin with nothing inside except this `napari.yaml`:

```yaml
name: napari-gruvbox
display_name: napari Gruvbox
contributions:
  themes:
  - id: gruvbox
    label: Gruvbox
    type: dark
    colors:
      canvas: black
      console: black
      background: '#282828'
      foreground: '#ebdbb2'
      primary: '#7c6f64'
      secondary: '#504945'
      highlight: '#98971a'
      text: '#ebdbb2'
      warning: '#fb4934'
      current: '#458588'
```

You can find it at https://github.com/brisvag/napari-gruvbox if you want to test.

**Comment by Czaki** [comment]
> Actually, I still can't seem to get my custom theme plugin to show up :/

when I develop this PR I have gruvbox in my options. Now I do not see it also. The scope of this PR is to preserve the selected themes between sessions. 

I will try to understand what happens and update this PR or open separate.

**Comment by Czaki** [comment]
Ok. It looks like I broke the theme discoverability in PR that absolutify all imports that have an impact on the order of imports. 

Now I added installing npe2 themes on create viewer, but I'm not sure if it is a proper way to do this (las it is another step performed in initialization that cannot be cached).

**Review by brisvag — APPROVED** [review]
Looks good to me, now it's working :) Should the `_install_npe2_themes` logic be happening inside `_initialize_plugins()`?

**Comment by Czaki** [comment]
> Looks good to me, now it's working :) Should the `_install_npe2_themes` logic be happening inside `_initialize_plugins()`?

Great idea. Done.

**Review by brisvag — APPROVED** [review]
Awesome!
